README.md
# Kopano API

Kopano API provides a web service with the endpoints to interface with Kopano
via HTTP APIs. The availability of APIs is controlled by plugins. See the [Plugins](#Plugins)
section below for details.

# Technologies

- Go

## Build dependencies

Make sure you have Go 1.13 or later installed. This project uses Go modules.

## Building from source

```
git clone <THIS-PROJECT> kapi
cd kapi
make
```

### Build with Docker

```
docker build -t kapid-builder -f Dockerfile.build .
docker run -it --rm -u $(id -u):$(id -g) -v $(pwd):/build kapid-builder
```

## Running Kopano API

```
KOPANO_GRAPI_SOCKETS=/run/kopano-grapi ./bin/kapid serve \
  --listen 127.0.0.1:8039 \
  --plugins-path=./plugins \
  --plugins=grapi \
  --iss=https://mykonnect.local
```

Where `--plugins-path` points to a folder containing Kopano API plugin modules.
Add environment variables as needed by those plugins. See next chapter for
more information about plugins.

The `--plugins` parameter can be used to select what plugins should be enabled.
It takes a comma separated value of plugin IDs as the plugin defined it during
its build time. If the `--plugins` parameter is empty (the default), all plugins
found will be activated.

The `--iss` parameter points to an OpenID Connect issuer with support for
discovery (Kopano Konnect). On start, the service will try discover OIDC details
and allow Bearer authentication with access tokens once successful. The `--iss`
parameter is mandatory.

## Plugins

Kopano API supports plugins to its behavior and ships with a bunch of
plugins to provide API endpoints from various data sources and different
purposes. An example plugin can be found in `plugins/example-plugin`.

### grapi: Kopano Groupware REST plugin (GRAPI)

Kopano API includes the plugin for Kopano Groupware REST. This plugin provides
access to Kopano Groupware RESTful API via `/api/gc/` URL routing prefix.

To specify where the grapi plugin can find its required GRAPI backend sockets,
specify the `KOPANO_GRAPI_SOCKETS` environment variable to point to the base
directory location. All `rest*.sock` files in that directory will be used as
upstream proxy paths for the REST api and all `notify*.sock` files in that
directory will be used as upstream proxy paths for the subscription socket API.

See the [grapi plugin README](https://stash.kopano.io/projects/KC/repos/kapi/browse/plugins/grapi/README.md) for further details.

### pubs: Kopano Pubsub and Webhook plugin

Kopano API includes a pub/sub system and webhook system via the Pubs plugin,
routed to the `/api/pubs` URL routing prefix. To specify the cryptographic
secret for the Pubs plugin use the environment variable
`KOPANO_PUBS_SECRET_KEY`. For more information on the Pubs plugin look at
'plugins/pubs/README.md'.

See the [pubs plugin README](https://stash.kopano.io/projects/KC/repos/kapi/browse/plugins/pubs/README.md) for further details.

### kvs: Kopano Key Value Store plugin

Kopano API inclues a key value store via the kvs plugin, routed to the
`/api/kvs` URL routing prefix. Kvs plugin needs configuration for its persistent
storage layer. Look at 'plugins/kvs/README.md' for more information.

See the [kvs plugin README](https://stash.kopano.io/projects/KC/repos/kapi/browse/plugins/kvs/README.md) for further details.

## Run unit tests

```
make test
```

## Testing the Kopano API

To test, some prerequisites are needed. A full fledged setup with TLS web server,
authentication provider and backend is strongly suggested. An quick way to set
this up is using a [Kopano Docker environment](https://github.com/kopano-dev/kopano-docker) which
provides:

  - TLS web server with [Kopano Web](https://stash.kopano.io/projects/KGOL/repos/kweb),
  - Authentication with [Kopano Konnect](https://stash.kopano.io/projects/KC/repos/konnect)
  - Backend with [GRAPI](https://stash.kopano.io/projects/KC/repos/grapi) and [Kopano Groupware](https://stash.kopano.io/projects/KC/repos/kopanocore)

Once you have all the bits in place and set up correctly, look at the `test`
folder in this project for a bunch of scripts and helpers to simplify testing
and give you ideas how to access the APIs provided by kapi.


CHANGELOG.md
# CHANGELOG

## Unreleased



## v0.15.1 (2020-10-27)

- Build with Go 1.14.10
- Add client secret support to test scripts
- Move Jenkins pipleline to warnings-ng plugin
- Fix compatibility with Go 1.15
- Build with Go 1.14.4


## v0.15.0 (2020-06-04)

- Improve example bench helpers
- Update 3rd party depdencies
- Update license ranger and generate 3rd party licenses from vendor folder
- Build with Go 1.14
- Add molotov for meet and calendar
- Fix a bunch of linter warnings
- Move plugins directly into daemon binary
- Update dependencies
- Update third party dependencies
- Update dependencies


## v0.14.0 (2019-11-25)

- Allow to start without any plugin
- Implement initialization order for plugins
- Bump linter to v1.21.0
- Update 3rd party modules
- Add minimal version unit test
- Use Go modules instead of Go dep


## v0.13.2 (2019-11-07)

- Build with Go 1.13.4


## v0.13.1 (2019-10-31)

- Build with Go 1.13.3
- Fix typo on README


## v0.13.0 (2019-10-01)

- Build with Go 1.13
- Add basic metrics
- Ensure BASE folder in fmt and check targets
- Improve README


## v0.12.2 (2019-09-30)

- Rebuild with Go 1.12.10


## v0.12.1 (2019-08-26)

- Update Go Dep to 0.5.4
- Compile on stretch to ensure compatibility


## v0.12.0 (2019-08-22)

- Use sticky backend selection for grapi
- Add GRAPI link to top level README
- Add more GRAPI information to README
- Unify env variable names of test scripts
- Update README
- Add instructions to example scripts
- Update plugin docs with additional details where those were missing


## v0.11.0 (2019-07-26)

- Add healthcheck sub command


## v0.10.0 (2019-05-17)

- Use correct Dep tag
- Use updated golint
- Build with Go 1.12 and Dep 0.5.2
- Update libkcoidc to 0.6.0 and use its new flexible logger
- Improve access token python example
- Fixup python syntax
- Add OAuth2 python3 example
- Add OIDC Python Kopano console helper
- Improve wording and fix typos
- oidc-client-example.html edited online with Bitbucket
- Show details and print OK for make check


## v0.9.0 (2019-01-28)

- Ensure third party caddy dependency is a compatible revision
- Require kopano/pubs scope to access pubs API
- Require kopano/kvs scope to access kvs
- Log what URL got access denied
- Fixup invalid debug log formatting


## v0.8.2 (2019-01-24)

- Show Go version in Jenkins


## v0.8.1 (2019-01-24)



## v0.8.0 (2019-01-22)

- Bump base copyright years to 2019
- Update dep docs and use pinned version
- Add CORS support to pubs (disabled by default)
- Add CORS support for kvs (disabled by default)
- Allow to set server to test CORS / use different host
- Update Makefile to no longer require local pubsjs
- scripts: remove duplicate ProtectSystem line
- Use latest released pubs version
- Update to updates Pubs module API


## v0.7.0 (2018-11-20)

- Update Caddy and its dependencies to same as used in kweb
- Add kvs batch mode for create or update
- Add code coverage reporting to Jenkins
- Add unit tests for KV
- Ensure array response format when recuse is requested
- Add docs for KVS
- Ensure to recurse when requested even if only one result
- Limit number of database connections to 10
- Add support for DELETE
- Add minimal DB open unit tests
- Add configuration for kvs
- Include db migrations in dist tarball
- Implement flexible persistent key value store
- Fixup deprecation warning in Jenkins
- Update license ranger to version with dep support
- Replace Glide with Go deps
- Build with Go 1.11
- Update build checks


## v0.6.4 (2018-09-26)

- Remove obsolete use of external environment files


## v0.6.3 (2018-09-21)

- Update libkcoidc to 0.4.3


## v0.6.2 (2018-09-06)

- Fixup cast to pointer


## v0.6.1 (2018-09-06)

- Update libkcoidc to 0.4.1


## v0.6.0 (2018-09-06)

- Use scope check from kcoidc
- Update libkcoidc to 0.4.0
- Refactor header injection so plugins can do it
- grapi: Rename groupware-core plugin to grapi
- Increase no-file limit to infinite


## v0.5.1 (2018-08-21)

- Add setup subcommand to binscript


## v0.5.0 (2018-08-17)

- Run Jenkins with Go 1.10
- Add log-level to config and avoid double timestamp for systemd
- Split of serve subcommand into extra file
- Add commandline args for log output control
- groupware-core: Include socket folder check in retry
- Add systemd unit with runner script and config
- groupware-core: Add README
- Error out when plugins-path is not found
- Subscripe to deleted in example
- examples: define a working example


## v0.4.2 (2018-08-10)

- Add purify hooks and debug
- Update Quill to version 2.0.0-dev.2 for table support
- Add HTML editor to example (Quill)


## v0.4.1 (2018-07-31)

- Add support for kcoidc debug log


## v0.4.0 (2018-07-18)

- Disable obsolete groupware-core v0 API endpoints
- Add scope validation to groupware-core and pubs


## v0.3.0 (2018-07-16)

- Update Kopano GC REST API support to v1
- Update pubjs in playground example


## v0.2.0 (2018-04-13)

- Implement injection of username header
- Add Makefile to example app
- Ignore test files
- Add helpers for testing
- Add docs for -iss parameter


## v0.1.0 (2018-03-19)

- Add OIDC discovery to validat access tokens
- Rename project to kapi for consistency
- pubs: Require minimal key size
- Fix format logging
- Add wrapper for hey for easy testing
- Only fetch realistic fields
- Add calendarView request to example webapp
- Add calendarView scenario
- Use pubjs in example
- Add 3rd party license information


## v0.0.1 (2018-02-14)

- Add Jenkinsfile
- example: Add pubs support to example webapp
- pubs: Add authentication for streams and logging
- Add auth submodule to support context base auth
- Update socket glob for KC changes
- Fixup: update README
- Add pubs plugin for minimal pubsub and webhooks
- Disable request HTTP log by default
- Add molotov test scenario
- gc: Implement subscriptions delegation
- gc: Only use mfr sockets for REST backend
- gc: Add support for CORS (disabled by default)
- Improve error behavior when no JSON is received

cmd/kapid/serve.go
/*
 * Copyright 2018 Kopano and its licensors
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License, version 3,
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

package main

import (
	"context"
	"crypto/tls"
	"fmt"
	"net"
	"net/http"
	_ "net/http/pprof"
	"net/url"
	"os"
	"runtime"
	"strings"
	"time"

	"github.com/prometheus/client_golang/prometheus/promhttp"
	"github.com/spf13/cobra"

	"stash.kopano.io/kc/kapi/server"
)

// Defaults.
const (
	defaultListenAddr = "127.0.0.1:8039"
)

func commandServe() *cobra.Command {
	serveCmd := &cobra.Command{
		Use:   "serve [...args]",
		Short: "Start server and listen for requests",
		Run: func(cmd *cobra.Command, args []string) {
			if err := serve(cmd, args); err != nil {
				fmt.Fprintf(os.Stderr, "Error: %v\n", err)
				os.Exit(1)
			}
		},
	}
	serveCmd.Flags().String("listen", defaultListenAddr, "TCP listen address")
	serveCmd.Flags().String("plugins-path", "", "Historic unused parameter")
	serveCmd.Flags().String("plugins", "", "Enabled plugin IDs. When empty, all found plugins are enabled. Separate multiple IDs with comma.")
	serveCmd.Flags().String("iss", "", "OIDC issuer URL")
	serveCmd.Flags().Bool("insecure", false, "Disable TLS certificate and hostname validation")
	serveCmd.Flags().Bool("log-timestamp", true, "Prefix each log line with timestamp")
	serveCmd.Flags().String("log-level", "info", "Log level (one of panic, fatal, error, warn, info or debug)")
	serveCmd.Flags().Bool("with-pprof", false, "With pprof enabled")
	serveCmd.Flags().String("pprof-listen", "127.0.0.1:6060", "TCP listen address for pprof")
	serveCmd.Flags().Bool("with-metrics", false, "Enable metrics")
	serveCmd.Flags().String("metrics-listen", "127.0.0.1:6039", "TCP listen address for metrics")

	return serveCmd
}

func serve(cmd *cobra.Command, args []string) error {
	ctx := context.Background()

	logTimestamp, _ := cmd.Flags().GetBool("log-timestamp")
	logLevel, _ := cmd.Flags().GetString("log-level")

	logger, err := newLogger(!logTimestamp, logLevel)
	if err != nil {
		return fmt.Errorf("failed to create logger: %v", err)
	}
	logger.Infoln("serve start")

	listenAddr, _ := cmd.Flags().GetString("listen")

	enabledPlugins := make([]string, 0)
	if pluginsString, strErr := cmd.Flags().GetString("plugins"); strErr == nil && pluginsString != "" {
		for _, id := range strings.Split(pluginsString, ",") {
			if id == "none" {
				enabledPlugins = nil
				break
			}
			enabledPlugins = append(enabledPlugins, strings.TrimSpace(id))
		}
	}
	if len(enabledPlugins) > 0 || enabledPlugins == nil {
		logger.Debugf("enabled plugins: %v", enabledPlugins)
	} else {
		logger.Debug("all plugins enabled")
	}

	var iss *url.URL
	if issString, parseErr := cmd.Flags().GetString("iss"); parseErr == nil && issString != "" {
		iss, parseErr = url.Parse(issString)
		if parseErr != nil {
			return fmt.Errorf("invalid iss url: %v", parseErr)
		}
	}
	if iss == nil {
		return fmt.Errorf("missing --iss parameter")
	}

	var tlsClientConfig *tls.Config
	tlsInsecureSkipVerify, _ := cmd.Flags().GetBool("insecure")
	if tlsInsecureSkipVerify {
		// NOTE(longsleep): This disable http2 client support. See https://github.com/golang/go/issues/14275 for reasons.
		tlsClientConfig = &tls.Config{
			InsecureSkipVerify: tlsInsecureSkipVerify,
		}
		logger.Warnln("insecure mode, TLS client connections are susceptible to man-in-the-middle attacks")
		logger.Debugln("http2 client support is disabled (insecure mode)")
	}
	client := &http.Client{
		Timeout: 60 * time.Second,
		Transport: &http.Transport{
			DialContext: (&net.Dialer{
				Timeout:   30 * time.Second,
				KeepAlive: 30 * time.Second,
				DualStack: true,
			}).DialContext,
			MaxIdleConns:          100,
			IdleConnTimeout:       90 * time.Second,
			TLSHandshakeTimeout:   10 * time.Second,
			ExpectContinueTimeout: 1 * time.Second,
			TLSClientConfig:       tlsClientConfig,
		},
	}

	// Metrics support.
	withMetrics, _ := cmd.Flags().GetBool("with-metrics")
	metricsListenAddr, _ := cmd.Flags().GetString("metrics-listen")
	if withMetrics && metricsListenAddr != "" {
		go func() {
			metricsListen := metricsListenAddr
			handler := http.NewServeMux()
			logger.WithField("listenAddr", metricsListen).Infoln("metrics enabled, starting listener")
			handler.Handle("/metrics", promhttp.Handler())
			listenErr := http.ListenAndServe(metricsListen, handler)
			if listenErr != nil {
				logger.WithError(listenErr).Errorln("unable to start metrics listener")
			}
		}()
	}

	srv, err := server.NewServer(listenAddr, "", iss, enabledPlugins, logger, client)
	if err != nil {
		return err
	}

	// Profiling support.
	withPprof, _ := cmd.Flags().GetBool("with-pprof")
	pprofListenAddr, _ := cmd.Flags().GetString("pprof-listen")
	if withPprof && pprofListenAddr != "" {
		runtime.SetMutexProfileFraction(5)
		go func() {
			pprofListen := pprofListenAddr
			logger.WithField("listenAddr", pprofListen).Infoln("pprof enabled, starting listener")
			err := http.ListenAndServe(pprofListen, nil)
			if err != nil {
				logger.WithError(err).Errorln("unable to start pprof listener")
			}
		}()
	}

	logger.Infof("serve started")
	return srv.Serve(ctx)
}
